wincode/src/io/mod.rs

```diff
@@ -0,0 +1,258 @@
+    ///
+    /// # Safety
+    ///
+    /// - `dst` must not overlap with the internal buffer.
+    #[inline]
+    fn copy_into_array<const N: usize>(
+        &mut self,
+        dst: &mut MaybeUninit<[u8; N]>,
+    ) -> ReadResult<()> {
+        let src = self.fill_array::<N>()?;
+        // SAFETY:
+        // - `fill_array` must do the appropriate bounds checking.
+        unsafe {
+            ptr::copy_nonoverlapping(src, dst.as_mut_ptr(), 1);
+            self.consume_unchecked(N);
+        }
+        Ok(())
+    }
+
+    /// Copy and consume exactly `size_of::<T>()` bytes from the [`Reader`] into `dst`.
+    ///
+    /// # Safety
+    ///
+    /// - `T` must be initialized by reads of `size_of::<T>()` bytes.
+    /// - `dst` must not overlap with the internal buffer.
+    #[inline]
+    unsafe fn copy_into_t<T>(&mut self, dst: &mut MaybeUninit<T>) -> ReadResult<()> {
+        let src = self.fill_exact(size_of::<T>())?;
+        // SAFETY:
+        // - `fill_exact` must do the appropriate bounds checking.
+        unsafe {
+            ptr::copy_nonoverlapping(src.as_ptr(), dst.as_mut_ptr().cast(), size_of::<T>());
+            self.consume_unchecked(size_of::<T>());
+        }
+        Ok(())
+    }
+
+    /// Copy and consume exactly `dst.len() * size_of::<T>()` bytes from the [`Reader`] into `dst`.
+    ///
+    /// # Safety
+    ///
+    /// - `T` must be initialized by reads of `size_of::<T>()` bytes.
+    /// - `dst` must not overlap with the internal buffer.
+    #[inline]
+    unsafe fn copy_into_slice_t<T>(&mut self, dst: &mut [MaybeUninit<T>]) -> ReadResult<()> {
+        let len = size_of_val(dst);
+        let bytes = self.fill_exact(len)?;
+        // SAFETY:
+        // - `fill_exact` must do the appropriate bounds checking.
+        unsafe {
+            ptr::copy_nonoverlapping(bytes.as_ptr(), dst.as_mut_ptr().cast(), len);
+            self.consume_unchecked(len);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum WriteError {
+    #[error("Attempting to write {0} bytes")]
+    WriteSizeLimit(usize),
+    #[cfg(feature = "std")]
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+#[cold]
+const fn write_size_limit(len: usize) -> WriteError {
+    WriteError::WriteSizeLimit(len)
+}
+
+pub type WriteResult<T> = core::result::Result<T, WriteError>;
+
+/// Trait for structured writing of bytes into a source of potentially uninitialized memory.
+pub trait Writer {
+    /// A variant of the [`Writer`] that can elide bounds checking within a given window.
+    type Trusted<'a>: Writer
+    where
+        Self: 'a;
+
+    /// Perform any cleanup or flushing of the writer.
+    ///
+    /// Be advised that this may be called on trusted [`Writer`] variants for subsections of the writer,
+    /// so make sure that those implementations don't perform any cleanup operations that would
+    /// invalidate the parent writer.
+    fn finish(&mut self) -> WriteResult<()> {
```

Comment:
it's a bit unclear to me when this has to be called or not. Some places have if
static { call it } else { dont }. I figured it was not necessary on Dynamic, but
some Dynamic branches are calling it?

elaborate? 😅